RequiredPropertyConstructorExtension doesn't order constructor fields correctly when OptionalAnnotations is used.

Given two Java classes
```java
public class TestSuper {
    private String requiredSuperField;
    @Nullable
    private String optionalSuperField;
}

public class TestSubClass extends TestSuper {
    private String requiredSubField;
}
```

The following output is given

```typescipt
export class TestSuper {
    readonly requiredSuperField: string;
    readonly optionalSuperField?: string;

    constructor(requiredSuperField: string, optionalSuperField?: string) {
        this.requiredSuperField = requiredSuperField;
        this.optionalSuperField = optionalSuperField;
    }
}

export class TestSubClass extends TestSuper {
    readonly requiredSubField: string;

    constructor(requiredSuperField: string, optionalSuperField?: string, requiredSubField: string) {
        super(requiredSuperField, optionalSuperField);
        this.requiredSubField = requiredSubField;
    }
}
```

This gives the following typescript compiler error:
error TS1016: A required parameter cannot follow an optional parameter. on the requiredSubField parameter to the constructor.

Maven settings.
```xml
ions>
    <extension>cz.habarta.typescript.generator.ext.RequiredPropertyConstructorExtension</extension>
</extensions>
<declarePropertiesAsReadOnly>true</declarePropertiesAsReadOnly>
<optionalAnnotations>
    <annotation>javax.annotation.Nullable</annotation>
</optionalAnnotations>
```

--- a/typescript-generator-core/src/main/java/cz/habarta/typescript/generator/ext/RequiredPropertyConstructorExtension.java
+++ b/typescript-generator-core/src/main/java/cz/habarta/typescript/generator/ext/RequiredPropertyConstructorExtension.java
@@ -92,6 +92,7 @@ public class RequiredPropertyConstructorExtension extends Extension {
     private static Optional<TsConstructorModel> createConstructor(TsBeanModel bean, TsModel model,
                                                                   Map<String, TsConstructorModel> generatedConstructors) {
         List<TsParameterModel> parameters = new ArrayList<>();
+        List<TsParameterModel> optionalParameters = new ArrayList<>();
         List<TsStatement> body = new ArrayList<>();
         TsType parent = bean.getParent();
         if (parent != null) {
@@ -107,7 +108,9 @@ public class RequiredPropertyConstructorExtension extends Extension {
             TsIdentifierReference[] callParameters = new TsIdentifierReference[parentParameters.size()];
             int i = 0;
             for (TsParameterModel parentParameter : parentParameters) {
-                parameters.add(parentParameter);
+                List<TsParameterModel> targetParameterList =
+                        parentParameter.tsType instanceof TsType.OptionalType ? optionalParameters : parameters;
+                targetParameterList.add(parentParameter);
                 callParameters[i] = new TsIdentifierReference(parentParameter.name);
                 i++;
             }
@@ -122,7 +125,10 @@ public class RequiredPropertyConstructorExtension extends Extension {
             if (predefinedValue.isPresent()) {
                 assignmentExpression = predefinedValue.get();
             } else {
-                parameters.add(new TsParameterModel(property.name, property.tsType));
+                TsParameterModel parameter = new TsParameterModel(property.name, property.tsType);
+                List<TsParameterModel> targetParameterList =
+                        property.tsType instanceof TsType.OptionalType ? optionalParameters : parameters;
+                targetParameterList.add(parameter);
                 assignmentExpression = new TsIdentifierReference(property.name);
             }
             TsMemberExpression leftHandSideExpression = new TsMemberExpression(new TsThisExpression(), property.name);
@@ -130,6 +136,7 @@ public class RequiredPropertyConstructorExtension extends Extension {
             TsExpressionStatement assignmentStatement = new TsExpressionStatement(assignment);
             body.add(assignmentStatement);
         }
+        parameters.addAll(optionalParameters);
         if(parameters.isEmpty() && body.isEmpty()) {
             return Optional.empty();
         }

--- a/typescript-generator-core/src/test/java/cz/habarta/typescript/generator/ext/RequiredPropertyConstructorExtensionTest.java
+++ b/typescript-generator-core/src/test/java/cz/habarta/typescript/generator/ext/RequiredPropertyConstructorExtensionTest.java
@@ -11,9 +11,10 @@ import cz.habarta.typescript.generator.TypeScriptGenerator;
 import cz.habarta.typescript.generator.TypeScriptOutputKind;
 import cz.habarta.typescript.generator.util.Utils;
 import java.lang.reflect.Type;
+import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Map;
-
+import javax.annotation.Nullable;
 import org.junit.Assert;
 import org.junit.Test;
 
@@ -60,6 +61,16 @@ public class RequiredPropertyConstructorExtensionTest {
 
     static class SecondClass extends SimpleClass {
         public int field3;
+    }
+
+    static class SimpleOptionalClass {
+        public String field1;
+        @Nullable
+        public Integer field2;
+    }
+
+    static class SecondOptionalClass extends SimpleOptionalClass {
+        public String field3;
     }
 
     @Test
@@ -120,6 +131,19 @@ public class RequiredPropertyConstructorExtensionTest {
         String result = generateTypeScript(settings, SecondClass.class);
 
         String expected = readResource("inheritance.ts");
+        Assert.assertEquals(expected, result);
+    }
+
+    @Test
+    public void testOptionalParameters() {
+        Settings settings = createBaseSettings();
+        settings.declarePropertiesAsReadOnly = true;
+        settings.optionalAnnotations = new ArrayList<>();
+        settings.optionalAnnotations.add(Nullable.class);
+
+        String result = generateTypeScript(settings, SecondOptionalClass.class);
+
+        String expected = readResource("optionalParameters.ts");
         Assert.assertEquals(expected, result);
     }
 

--- a/typescript-generator-core/src/test/resources/ext/RequiredPropertyConstructorExtensionTest-optionalParameters.ts
+++ b/typescript-generator-core/src/test/resources/ext/RequiredPropertyConstructorExtensionTest-optionalParameters.ts
@@ -1,0 +1,20 @@
+/* tslint:disable */
+
+export class SimpleOptionalClass {
+    readonly field1: string;
+    readonly field2?: number;
+
+    constructor(field1: string, field2?: number) {
+        this.field1 = field1;
+        this.field2 = field2;
+    }
+}
+
+export class SecondOptionalClass extends SimpleOptionalClass {
+    readonly field3: string;
+
+    constructor(field1: string, field3: string, field2?: number) {
+        super(field1, field2);
+        this.field3 = field3;
+    }
+}

